pkg/kubeclient/kubelet_client.go

```diff
@@ -150,59 +161,89 @@ func (client *KubeletClient) GetSummary(host string) (*stats.Summary, error) {
 	} else {
 		entry := &CacheEntry{
 			statsSummary: summary,
-			machineInfo:  nil,
 			used:         false,
 		}
 		client.cache[host] = entry
 	}
 	return summary, err
 }
 
-func (client *KubeletClient) GetMachineInfo(host string) (*cadvisorapi.MachineInfo, error) {
-	// Get the data
-	var minfo cadvisorapi.MachineInfo
-	err := client.ExecuteRequestAndGetValue(host, specPath, &minfo)
-	// Lock and check cache
+// get node single-core Frequency, in MHz
+func (client *KubeletClient) GetNodeCpuFrequency(node *v1.Node) (float64, error) {
+	ip, err := util.GetNodeIPForMonitor(node, types.KubeletSource)
+	if err != nil {
+		glog.Errorf("Failed to IP for node %s: %s", node.Name, err)
+		return 0, err
+	}
+
+	// Get value from cache if exists
 	client.cacheLock.Lock()
-	defer client.cacheLock.Unlock()
-	entry, entryPresent := client.cache[host]
+	entry, entryPresent := client.cache[ip]
+	if entryPresent {
+		if entry.nodeCpuFreq != nil {
+			nodeFreq := *entry.nodeCpuFreq
+			client.cacheLock.Unlock()
+			return nodeFreq, nil
+		}
+	}
+	client.cacheLock.Unlock()
+
+	// We could not get a valid cpu frequency from cache, discover it.
+	var nodeFreq float64
+	minfo, err := client.GetMachineInfo(ip)
+	freqDiscovered := true
 	if err != nil {
-		if entryPresent {
-			entry.used = true
-			if entry.machineInfo == nil {
-				glog.V(2).Infof("unable to retrieve machine[%s] machine info: %v. The cached value unavailable", host, err)
-				return nil, err
+		glog.Errorf("Failed to get machine[%s] cpu.frequency from kubelet: %v. Will try pod based getter.", node.Name, err)
+		nodeIsActive := util.NodeIsReady(node) && util.NodeIsSchedulable(node)
+		if nodeIsActive {
+			nodeFreq, err = client.fallbkCpuFreqGetter.GetFrequency(node.Name)
+			if err != nil {
+				glog.Errorf("Failed to get cpufreq from getter %s: %s. Default value will be used.", node.Name, err)
+				freqDiscovered = false
 			}
-			glog.V(2).Infof("unable to retrieve machine[%s] machine info: %v. Using cached value", host, err)
-			return entry.machineInfo, nil
 		} else {
-			glog.Errorf("failed to get machine[%s] machine info: %v. No cache available", host, err)
-			return &minfo, err
+			freqDiscovered = false
 		}
+	} else {
+		nodeFreq = util.MetricKiloToMega(float64(minfo.CpuFrequency))
 	}
-	// Fill in the cache
+
+	if !freqDiscovered {
+		// We probably could not discover the cpufreq by any means and default
+		// value is what we need to stick to. However, we might be able to discover
+		// a valid one in the next discovery cycle.
+
+		return client.defaultCpuFreq, nil
```

Comment:
So just to confirm, in case we cannot discover cpufreq from any node, we still use the `DefaultCpuFreq`, right? If so, as @enlinxu has suggested, this will overwrite the used and capacity value of a node with the fake value in a stitched environment. In this situation, I still think we should fail the discovery (or the validation). Sending no value is still better than sending wrong values.

Reply:
The issue of patching with wrong default in a stitched env is there ( I did verify it and the infra values are overridden by the ones sent from kubeturbo).

Having said that; I also had a chat with @esara before posting this yesterday. His point of view is that this is most likely going to be used for demo purposes in the demo envs (probably until we land a proper milicore based solution). Its less likely that we would have a real customer using 1.18+ cluster already. For demo envs, its also more likely that there are transient issues when we start off (for example image not available internally), where if we fail discovery we wont have anything to show in supply chain. It makes sense to show something, rather then nothing in such cases.

I have updated the code to catch all possible issues and they will be Error logged anyways.

If it might make more sense, I can also updated the default cpu freq to `2600 MHz` (in place of `2000`), which is what I have seen more often (2.6/2.7 GHz) across clusters.

Reply:
Thanks for the explanation. Make senses for demo envs.